packages/fast-usdc/src/exos/liquidity-pool.js

```diff
@@ -0,0 +1,232 @@
+ * @import {MakeRecorderKit} from '@agoric/zoe/src/contractSupport/recorder.js'
+ * @import {USDCProposalShapes, ShareWorth} from '../pool-share-math.js'
+ */
+
+const { add, isEqual } = AmountMath;
+
+/** @param {Brand} brand */
+const makeDust = brand => AmountMath.make(brand, 1n);
+
+/**
+ *
+ * @param {ZCFSeat} poolSeat
+ * @param {ShareWorth} shareWorth
+ * @param {Brand} USDC
+ */
+const checkPoolBalance = (poolSeat, shareWorth, USDC) => {
+  const available = poolSeat.getAmountAllocated('USDC', USDC);
+  const dust = makeDust(USDC);
+  isEqual(add(available, dust), shareWorth.numerator) ||
+    Fail`🚨 pool balance ${q(available)} inconsistent with shareWorth ${q(shareWorth)}`;
+};
+
+/**
+ * @param {Zone} zone
+ * @param {ZCF} zcf
+ * @param {Brand<'nat'>} USDC
+ * @param {{
+ *   makeRecorderKit: MakeRecorderKit;
+ * }} tools
+ */
+export const prepareLiquidityPoolKit = (zone, zcf, USDC, tools) => {
+  return zone.exoClassKit(
+    'Fast Liquidity Pool',
+    {
+      feeSink: M.interface('feeSink', {
+        receive: M.call(AmountShape, PaymentShape).returns(M.promise()),
+      }),
+      external: M.interface('external', {
+        publishShareWorth: M.call().returns(M.promise()),
+      }),
+      depositHandler: M.interface('depositHandler', {
+        handle: M.call(SeatShape, M.any()).returns(M.promise()),
+      }),
+      withdrawHandler: M.interface('withdrawHandler', {
+        handle: M.call(SeatShape, M.any()).returns(M.promise()),
+      }),
+      public: M.interface('public', {
+        makeDepositInvitation: M.call().returns(M.promise()),
+        makeWithdrawInvitation: M.call().returns(M.promise()),
+        getPublicTopics: M.call().returns(TopicsRecordShape),
+      }),
+    },
+    /**
+     * @param {ZCFMint<'nat'>} shareMint
+     * @param {StorageNode} node
+     */
+    (shareMint, node) => {
+      const { brand: PoolShares } = shareMint.getIssuerRecord();
+      const proposalShapes = makeProposalShapes({ USDC, PoolShares });
+      const shareWorth = makeParity(makeDust(USDC), PoolShares);
+      const { zcfSeat: poolSeat } = zcf.makeEmptySeatKit();
+      const shareWorthRecorderKit = tools.makeRecorderKit(
+        node,
+        /** @type {TypedPattern<Ratio>} */ (RatioShape),
+      );
+      return {
+        shareMint,
+        shareWorth,
+        poolSeat,
+        PoolShares,
+        proposalShapes,
+        shareWorthRecorderKit,
+      };
+    },
+    {
+      feeSink: {
+        /**
+         * @param {Amount<'nat'>} amount
+         * @param {Payment<'nat'>} payment
+         */
+        async receive(amount, payment) {
+          const { poolSeat, shareWorth } = this.state;
+          const { external } = this.facets;
+          await depositToSeat(
+            zcf,
+            poolSeat,
+            harden({ USDC: amount }),
+            harden({ USDC: payment }),
+          );
+          this.state.shareWorth = withFees(shareWorth, amount);
+          await external.publishShareWorth();
```

Comment:
if the publishing fails you want `receive` to throw?

I don't know of a way to recover from failed publishing so I tend to make the publish methods sync (and `void` their `.write()`)

Another reason to await is to get a signal that the write happened, but the only consumers are off-chain so I don't see why the contract should wait.

Reply:
my brain is leaky. I meant to head that way but forgot to or something